src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -2506,9 +2506,20 @@ public DynamicObject upcase(DynamicObject string,
             }
         }
 
-        @TruffleBoundary
-        private boolean singleByteUpcase(byte[] bytes, int s, int end) {
-            return StringSupport.singleByteUpcase(bytes, s, end);
+        private boolean singleByteUpcase(byte[] bytes) {
+            boolean modify = false;
+
+            for (int i = 0; i < bytes.length; i++) {
+                final byte b = bytes[i];
+
+                // Check if between 'a' - 'z'.
+                if (b >= 0x61 && b <= 0x7a) {
+                    bytes[i] ^= 0x20;
```

Comment:
Why the xor and not just `+` here?

Reply:
Bah. I used to have lower and upper bounds as passed variables and the `^` worked in both cases. I had to break them out for `swapcase`. I'll fix.

Reply:
Yeah I saw it's a nice trick as it works in both cases. I guess it doesn't matter in practice, so this is fine.